waspc/src/Wasp/Generator/ServerGenerator/OperationsG.hs

```diff
@@ -10,45 +10,95 @@ where
+      genActionTypesFile spec
+    ]
+    <++> genQueries spec
+    <++> genActions spec
 
 genQueries :: AppSpec -> Generator [FileDraft]
-genQueries spec = mapM (genQuery spec) (AS.getQueries spec)
+genQueries = mapM genQuery . AS.getQueries
 
 genActions :: AppSpec -> Generator [FileDraft]
-genActions spec = mapM (genAction spec) (AS.getActions spec)
+genActions = mapM genAction . AS.getActions
+
+genQueryTypesFile :: AppSpec -> Generator FileDraft
+genQueryTypesFile spec = genOperationTypesFile tmplFile dstFile operations isAuthEnabledGlobally
+  where
+    tmplFile = [relfile|src/queries/types.ts|]
+    dstFile = [relfile|src/queries/types.ts|]
+    operations = map (uncurry AS.Operation.QueryOp) $ AS.getQueries spec
+    isAuthEnabledGlobally = isAuthEnabled spec
+
+genActionTypesFile :: AppSpec -> Generator FileDraft
+genActionTypesFile spec = genOperationTypesFile tmplFile dstFile operations isAuthEnabledGlobally
+  where
+    tmplFile = [relfile|src/actions/types.ts|]
+    dstFile = [relfile|src/actions/types.ts|]
+    operations = map (uncurry AS.Operation.ActionOp) $ AS.getActions spec
+    isAuthEnabledGlobally = isAuthEnabled spec
 
 -- | Here we generate JS file that basically imports JS query function provided by user,
 --   decorates it (mostly injects stuff into it) and exports. Idea is that the rest of the server,
 --   and user also, should use this new JS function, and not the old one directly.
-genQuery :: AppSpec -> (String, AS.Query.Query) -> Generator FileDraft
-genQuery _ (queryName, query) = return $ C.mkTmplFdWithDstAndData tmplFile dstFile (Just tmplData)
+genQuery :: (String, AS.Query.Query) -> Generator FileDraft
+genQuery (queryName, query) = return $ C.mkTmplFdWithDstAndData tmplFile dstFile (Just tmplData)
   where
     operation = AS.Operation.QueryOp queryName query
     tmplFile = C.asTmplFile [relfile|src/queries/_query.ts|]
     dstFile = C.serverSrcDirInServerRootDir </> queryFileInSrcDir queryName
     tmplData = operationTmplData operation
 
+genOperationTypesFile ::
+  Path' (Rel C.ServerTemplatesDir) File' ->
+  Path' (Rel C.ServerRootDir) File' ->
+  [AS.Operation.Operation] ->
+  Bool ->
+  Generator FileDraft
+genOperationTypesFile tmplFile dstFile operations isAuthEnabledGlobally =
+  return $ C.mkTmplFdWithDstAndData tmplFile dstFile (Just tmplData)
+  where
+    tmplData =
+      object
+        [ "operations" .= map operationTypeData operations,
+          "allUseAuth" .= all usesAuth operations,
+          "someUseAuth" .= any usesAuth operations,
+          "allEntities" .= nub (concatMap entityNames operations)
+        ]
+    operationTypeData operation =
+      object
+        [ "typeName" .= toUpperFirst (getName operation),
+          "entities" .= entityNames operation,
+          "usesAuth" .= usesAuth operation
+        ]
+    entityNames = maybe [] (map AS.refName) . AS.Operation.getEntities
+    usesAuth = fromMaybe isAuthEnabledGlobally . AS.Operation.getAuth
```

Comment:
Maybe make this produce value immediately 
```
fromMaybe isAuthEnabledGlobally . AS.Operation.getAuth $ operations
```
so you can use it like `any usesAuth` and `all usesAuth`.

Reply:
The thing is, any and all actually take a predicate function which they run on the given list(Foldable). So what you wrote couldn't be done. For that thought, you could use functions like `and` and `or`.

Reply:
I'd need two values (the `usesAuth` function, and the result of mapping it over the operations that I would then use `and` and `or`), which I think makes the entire a little less readable (you need three "jumps" to understand what's going on).
It is theoretically more efficient, but these arrays are almost always smaller than 10, and always in the two digit space.
Anyway, I'd like to leave it like this if that's fine with you.